### DISCUSSION TITLE
DSM Forward Programming

### DISCUSSION THREAD

**Discussion by Dweinbach** [pascallanger/DIY-Multiprotocol-TX-Module#748] [General]
Hi:

Is/are there any "Forward Programming" guru(s) here? 

The current state of the DSM forward programming script is definitely not ready for prime time - as in cannot replace a Spectrum TX. 

The most obvious reason is that there are many, many strings that are wrong or missing, and some workflows and logic that are incorrect as well.

But a more daunting (to me at least) concern is the "Relearn servo settings" and "initial safe setup" functions, which I believe (but cannot prove) are handed a packet of channel config settings front the TX model config, and these are used to set IMPORTANT things like servo reversing. Does the current Forward Programming script take this into account, and if so how do we know it is doing it correctly?

I was going to put some work into correcting some strings and identifying some of the "business logic" issues, but without confidence in the servo settings learning, it seems pretty moot, since you need to do a lot of work int he Spektrum TX anyway.

Just curious what is known about this aspect of forward programming, and how far along the script is in this particular regard?

Dave

**Comment by frankiearzu** [comment]
Dave, 
Im working on making the DSM forward programming more usefull.  Took the original code from Pascal, and added a lot of features to make it easier to support.  
1. Separate the GUI from the DSM Fwd Prog logic, so that the UI can be develop independently (created an RX AR631 simulator).
2. two GUI, one color/touchscreen support, one Black/White that could be adapted easilly for smaller screens (same shared DSM Fwd Prog Logic)
3. Log the conversation between the RX/TX in a log file translating the messages into English type of message types and parameters.
4. on the DSM Fwd Prog Engine, added extra metadata to the messages so that you can show the messages in BOLD, etc.  Also added list of valid options to some List menus when the values are not contiguos.. that eliminates showing invalid values.

I think my new code is stable enough to check it in.. Will do that so that more people can play with it.

![Capture1](https://user-images.githubusercontent.com/32604366/201715239-9a02319f-5fe0-42c2-bbd3-487b857b86aa.PNG)
![Capture2](https://user-images.githubusercontent.com/32604366/201715262-64da8ae6-5f47-4d6c-9383-3bfd15c34585.PNG)
![Capture3](https://user-images.githubusercontent.com/32604366/201715907-534082ea-2fb3-4f59-947c-eda67b533f5b.PNG)

**Reply by Dweinbach** [reply]
Wow. Great news. It seems the feature has languished, and without it it is really not possible to completely move away from Spektrum TX's to openTX or EdgeTX.

Any thoughts about my "Relearn servo settings" question? is there a way to "sniff" what a spektrum TX sends to the RX when that command is sent? I assume not. But then how do we know what to send?

**Reply by frankiearzu** [reply]
Was trying to send you a message directly.. but still don't find how :-)
Will try to check in my code so that we can progress with more tools to enhance this as a group

I sent an email to Pascal when i started this improvements about 1m ago of how did the "Capture" the conversation between RX/TX to start reverse engineering.. Did not get a response. But there has to be a way to "Capture" the convesation.  That will help compare what is happening with a Real Spektrum product. Will ask Pascal again.

Some screens has to send more messages to the RX than just UI navigation.. Was looking at some screens who has to reset/re-boot the RX.. the screens render in the GUI, but they don't reboot the RX. Has to be some other message that we have to send while navigating the Menus.

Specifically on that menu (**Servo Realm**), the RX/TX gets in a loop processing lines who has not fully been resverse engineeres.. 
In the code I document it as  "Unknown_Lines".   Pascal figured out how to skip them when we receive them in the inital page of the receiver, but on other screens, they are still not reversed engineered.  Look at the end of the log.

============================  Servo Realm (gets stuck in the Unknown Lines) ========================

> 17.980 MENU_LINES: DSM_GotoMenu(0x1022)
17.980 MENU_TITLE: SEND DSM_getMenu(MenuId=0x1022 StartLine=0)
18.070 MENU_TITLE: RESPONSE Menu: M[Id=0x1022 P=0x0 N=0x0 B=0x1010 Text="System Setup"[0x86]]
18.130 MENU_LINES: SEND DSM_getFirstMenuLine(MenuId=0x1022)
18.220 MENU_LINES: RESPONSE MenuLine: L[#0 T=M VId=0x1023 val=nil [0->0,3] Text="Relearn Servo Settings"[0x190]  MId=0x1022 ]
18.280 MENU_LINES: SEND DSM_getNextLine(MenuId=0x1022,LastLine=0)
18.390 MENU_LINES: RESPONSE MenuLine: L[#1 T=M VId=0x1025 val=nil [0->0,2] Text="Orientation"[0x80]  MId=0x1022 ]
18.430 MENU_LINES: SEND DSM_getNextLine(MenuId=0x1022,LastLine=1)
18.540 MENU_LINES: RESPONSE MenuLine: L[#2 T=M VId=0x1029 val=nil [0->0,2] Text="Gain Channel Select"[0xAD]  MId=0x1022 ]
18.580 MENU_LINES: SEND DSM_getNextLine(MenuId=0x1022,LastLine=2)
18.700 MENU_LINES: RESPONSE MenuLine: L[#3 T=M VId=0x102A val=nil [0->0,2] Text="SAFE/Panic Mode Setup"[0xCA]  MId=0x1022 ]
18.750 MENU_LINES: SEND DSM_getNextLine(MenuId=0x1022,LastLine=3)
18.850 MENU_LINES: RESPONSE MenuLine: L[#4 T=M VId=0x1032 val=nil [0->0,2] Text="Utilities"[0x240]  MId=0x1022 ]
18.900 MENU_LINES: SEND DSM_getNextLine(MenuId=0x1022,LastLine=4)
MENU_LINES: EVT_TOUCH_TAP 144,73
MENU_LINES: EVT_TOUCH_TAP 144,73
MENU_LINES: EVT_VIRTUAL_ENTER,  SelLine=0
19.930 MENU_LINES: DSM_GotoMenu(0x1023)
19.930 MENU_TITLE: SEND DSM_getMenu(MenuId=0x1023 StartLine=0)
20.040 MENU_TITLE: RESPONSE Menu: M[Id=0x1023 P=0x0 N=0x0 B=0x1022 Text="Relearn Servo Settings"[0x190]]
20.070 MENU_LINES: SEND DSM_getFirstMenuLine(MenuId=0x1023)
**20.130 MENU_LINES: RESPONSE MenuUknownLine_0x05: LineNum=0  DATA=RX: 09 05 00 1F 00 00 00 07 00 00 00 00 00 00 00 00
20.190 MENU_UNKNOWN_LINES: CALL DSM_getNextUknownLine_0x05(LastLine=0)
20.640 MENU_UNKNOWN_LINES: RESPONSE MenuUknownLine_0x05: LineNum=0  DATA=RX: 09 05 00 1F 00 00 00 07 00 00 00 00 00 00 00 00
ERROR: Received Same menu line, exiting the loop to prevent disconnect
**
20.710 WAIT_CMD: SEND DSM_sendHeartbeat()
21.150 WAIT_CMD: RESPONSE MenuUknownLine_0x05: LineNum=0  DATA=RX: 09 05 00 1F 00 00 00 07 00 00 00 00 00 00 00 00
ERROR: Received Same menu line, exiting the loop to prevent disconnect_**

===================  Original First menu (Passed the unknow lines OK) ===========================
>0.000 RX_VERSION: SEND DSM_getRxVersion()
0.090 RX_VERSION: RESPONSE Receiver=AR631 Version 2.38.5
0.130 MENU_TITLE: SEND DSM_getMainMenu()
0.240 MENU_TITLE: RESPONSE Menu: M[Id=0x1000 P=0x0 N=0x0 B=0x0 Text="Main Menu"[0x4B]]
0.280 MENU_LINES: SEND DSM_getFirstMenuLine(MenuId=0x1000)
0.390 MENU_LINES: RESPONSE MenuUknownLine_0x05: LineNum=0  DATA=RX: 09 05 00 01 00 00 00 07 00 00 00 00 00 00 00 00
**0.430 MENU_UNKNOWN_LINES: CALL DSM_getNextUknownLine_0x05(LastLine=0)
0.490 MENU_UNKNOWN_LINES: RESPONSE MenuUknownLine_0x05: LineNum=1  DATA=RX: 09 05 01 01 00 00 00 07 00 00 00 00 00 00 00 00
0.540 MENU_UNKNOWN_LINES: CALL DSM_getNextUknownLine_0x05(LastLine=1)
0.640 MENU_UNKNOWN_LINES: RESPONSE MenuUknownLine_0x05: LineNum=2  DATA=RX: 09 05 02 01 00 00 00 07 00 00 00 00 00 00 00 00**
0.690 MENU_UNKNOWN_LINES: CALL DSM_getNextUknownLine_0x05(LastLine=2)
0.750 MENU_UNKNOWN_LINES: RESPONSE MenuUknownLine_0x05: LineNum=3  DATA=RX: 09 05 03 01 00 00 00 07 00 00 00 00 00 00 00 00
0.790 MENU_UNKNOWN_LINES: CALL DSM_getNextUknownLine_0x05(LastLine=3)
0.900 MENU_UNKNOWN_LINES: RESPONSE MenuUknownLine_0x05: LineNum=4  DATA=RX: 09 05 04 01 00 00 00 07 00 00 00 00 00 00 00 00
0.940 MENU_UNKNOWN_LINES: CALL DSM_getNextUknownLine_0x05(LastLine=4)
1.050 MENU_UNKNOWN_LINES: RESPONSE MenuUknownLine_0x05: LineNum=5  DATA=RX: 09 05 05 01 00 00 00 07 00 00 00 00 00 00 00 00
1.110 MENU_UNKNOWN_LINES: CALL DSM_getNextUknownLine_0x05(LastLine=5)
1.160 MENU_UNKNOWN_LINES: RESPONSE MenuLine: L[#0 T=M VId=0x1010 val=nil [0->0,3] Text="Gyro settings"[0xF9]  MId=0x1000 ]
1.200 MENU_LINES: SEND DSM_getNextLine(MenuId=0x1000,LastLine=0)
1.320 MENU_LINES: RESPONSE MenuLine: L[#1 T=M VId=0x105E val=nil [0->0,2] Text="Other settings"[0x227]  MId=0x1000 ]
1.350 MENU_LINES: SEND DSM_getNextLine(MenuId=0x1000,LastLine=1)_

**Reply by pascallanger** [reply]
Sorry I must have missed your email.
I can dump both ways:
 - By emulating a Spektrum RX with forward programming using Multi and therefore get the commands/responses from on original Spektrum radio : with that you could dump all the texts, pictures,... or see how the TX reacts to the RX commands/responses. I've posted on the forward programming opened issue.
 - Dump directly from the radio Spektrum RF module using a cheap signal analyzer to get the packets sent and received. Just need to solder a few wires to the RF module and convert the data.

**Reply by Dweinbach** [reply]
> Sorry I must have missed your email. I can dump both ways:
...
> * Dump directly from the radio Spektrum RF module using a cheap signal analyzer to get the packets sent and received. Just need to solder a few wires to the RF module and convert the data.

It seems like this is the missing information - what does the TX send the RX with the "relearn servo settings" command. This and the reboot/restart functions are the most obvious (but not necessarily the only) gaps. I don't have the programming knowledge to implement the solution, but if you could do a couple captures from the spektrum TX module with a logic analyzer with 2 different data sets (change servo trims, centering, travel, invert, etc between captures), then we could probably reverse-engineer what needs to be sent. @pascallanger
